src/analyzer.ts

```diff
@@ -8,762 +8,139 @@
-        });
-      });
-      return Promise.all(depPromises);
-    });
-  };
-
-  /**
-   * List all the html dependencies for the document at `href`.
-   * @param  {string}                   href      The href to get dependencies for.
-   * @param  {Object.<string,boolean>=} found     An object keyed by URL of the
-   *     already resolved dependencies.
-   * @param  {boolean=}                transitive Whether to load transitive
-   *     dependencies. Defaults to true.
-   * @return {Array.<string>}  A list of all the html dependencies.
-   */
-  _getDependencies(href:string, found?:{[url:string]: boolean}, transitive?:boolean):Promise<string[]> {
-    if (found === undefined) {
-      found = {};
-      found[href] = true;
-    }
-    if (transitive === undefined) {
-      transitive = true;
-    }
-    var deps: string[] = [];
-    return this.load(href).then((htmlMonomer) => {
-      var transitiveDeps: Promise<string[]>[] = [];
-      htmlMonomer.depHrefs.forEach((depHref) => {
-        if (found[depHref]) {
-          return;
-        }
-        deps.push(depHref);
-        found[depHref] = true;
-        if (transitive) {
-          transitiveDeps.push(this._getDependencies(depHref, found));
-        }
-      });
-      return Promise.all(transitiveDeps);
-    }).then(function(transitiveDeps) {
-      var alldeps = transitiveDeps.reduce(function(a, b) {
-        return a.concat(b);
-      }, []).concat(deps);
-      return alldeps;
-    });
-  };
-
-  /**
-   * Returns the elements defined in the folder containing `href`.
-   * @param {string} href path to search.
-   */
-  elementsForFolder(href: string): ElementDescriptor[] {
-    return this.elements.filter(function(element){
-      return matchesDocumentFolder(element, href);
-    });
-  };
-
-  /**
-   * Returns the behaviors defined in the folder containing `href`.
-   * @param {string} href path to search.
-   * @return {Array.<BehaviorDescriptor>}
-   */
-  behaviorsForFolder(href:string):BehaviorDescriptor[] {
-    return this.behaviors.filter(function(behavior){
-      return matchesDocumentFolder(behavior, href);
-    });
-  };
-
-  /**
-   * Returns a promise that resolves to a POJO representation of the import
-   * tree, in a format that maintains the ordering of the HTML imports spec.
-   * @param {string} href the import to get metadata for.
-   * @return {Promise}
-   */
-  metadataTree(href:string) {
-    return this.load(href).then((monomer) =>{
-      var loadedHrefs: {[href: string]: boolean} = {};
-      loadedHrefs[href] = true;
-      return this._metadataTree(monomer, loadedHrefs);
-    });
-  };
+   * @param {string} url the location of the file to analyze
+   * @return {Promise<DocumentDescriptor>}
+   */
+  async analyze(url: string): Promise<DocumentDescriptor> {
+    if (this._documentDescriptors.has(url)) {
+      return this._documentDescriptors.get(url);
+    }
+    let promise = (async () => {
+      let document = await this.load(url);
```

Comment:
We want to set the promise just for the loading and analyzing into `this._documentDescriptors`, so no.